### PR TITLE
[Feature:PHP] Show times in user's timezone site wide

### LINKED ISSUE #5420: Site-wide time zone handling
This issue is the successor to PR #5407

I'd like to finish implementing proper time zone handling site wide.  Below is a check list of all spots on the site which display date or time.  Unless otherwise specified all these data/times will be changed over to displaying the date/time adjusted for the user's time zone.  If the user hasn't set a time zone then the time will be formatted according to the server's time zone.  This isn't just limited to display of dates.  Collecting date/time information from an instructor should adjust the date/time stamp from the user's time zone into the server time zone before saving to the database/file.

- [x] Buttons on the Gradeables navigation page
- [x] Times on the gradeable submission page
- [x] Lab grading with 'Show Dates Graded' selected
- [x] Dates on the 'Dates' tab of the edit gradeable page
- [x] Dates appear inside the TA grading interface
- [x] Notifications
- [x] Office hours queue
- [x] Course material release times
- [x] Discussion forum list of threads
- [x] Discussion forum viewing a post
- [x] Discussion forum create thread lock date
- [x] Latedays allowed
- [x] Excused Absence Extensions
- [x] Grade Reports - time under grade summaries button
- [x] My Late Days/Extensions
- [x] Grade Inquiries
- [x] Peer Grading

### PR DESCRIPTION
**Pull request by Kevin-MM**
Closes #5420

### What is the current behavior?
Some date/times on the site are converted to the logged in user's time zone.  Others are converted to server time zone.

### What is the new behavior?
All date/times on the site should now be displayed in the logged in user's time zone.  If the user has not set a time zone then date/times will be converted to the server's time zone.

The one exception to this currently is the `Excused absence extensions` page which is currently undergoing other bug fixes.  Time zone handling for that page will be addressed in a later PR.

### Other information?
**Testing Tip:** To avoid confusion when inputting date/times using a date time picker, ensure your submitty time zone *AND* computer time zone match.

### CODE REVIEW

File: site/app/models/Config.php [hunk 1]
```diff
@@ -258,6 +261,10 @@ class Config extends AbstractModel {
     public function __construct(Core $core) {
         parent::__construct($core);
         $this->timezone = new \DateTimeZone($this->default_timezone);
+
+        // For now this will be set to 'MDY', and configured as a property of the Config class
+        // Eventually this should be moved to the User class and configured on a per-user basis
+        $this->date_time_format = new DateTimeFormat($this->core, 'MDY');
```
Comment (MasterOdin) [inline site/app/models/Config.php:267]: Is there buy-in from @bmcutler that this is something we're going to support in the proposed fashion at some point in the future?

**Comment by MasterOdin** [comment]
Also, please name your PRs per the [How to Contribute](https://submitty.org/developer/how_to_contribute#how-to-make-a-pull-request-pr-to-submitty) doc (I would probably go with `[Feature:PHP] Show times in user's timezone sitewide` or something). Want to make sure all primary devs of Submitty are following our own set policies.

File: site/tests/app/libraries/DateUtilsTester.php [hunk 2]
```diff
@@ -137,6 +137,95 @@ public function testDateTimeToStringNoOffset() {
...
+        $time_zone = new \DateTimeZone('America/Los_Angeles');
+
+        $time_stamp = '2020-12-01 12:00:00-05';
+        $date_time = DateUtils::parseDateTime($time_stamp, $time_zone);
+
+        $expected_string = '2020-12-01 09:00:00';
+        $captured_string = $date_time->format($format);
+
+        $this->assertEquals($expected_string, $captured_string);
+    }
```
[the remark this answers is not in the document]
Comment (Kevin-MM) [reply site/tests/app/libraries/DateUtilsTester.php:226]: The time stamps in the tests I added are formatted like the time stamps that the database returns.  The time stamps in the other tests are not.  Also I wanted to test converting them into real time zones that undergo day light savings time.

**Comment by jchulton** [comment]
I've tested this with everything I can think of: course materials, gradeables (late days, TA_grading view), office hours queue, discussion forum,  grade inquiries. It all seems to work

File: site/tests/app/libraries/DateUtilsTester.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (MasterOdin) [reply site/tests/app/libraries/DateUtilsTester.php:226]: ah, alright, makes sense. We can condense these, with the two tests above with [parameterized tests](https://submitty.org/developer/testing/php_unit_tests#parameterized-php-unit-tests).

```diff
index f310fd47d..861b5aeca 100644
--- a/site/tests/app/libraries/DateUtilsTester.php
+++ b/site/tests/app/libraries/DateUtilsTester.php
@@ -95,10 +95,26 @@ class DateUtilsTester extends \PHPUnit\Framework\TestCase {
         DateUtils::parseDateTime('this is an invalid date', new \DateTimeZone('America/New_York'));
     }

-    public function testParseDateTimeString() {
-        $timezone = new \DateTimeZone("Etc/GMT+5");
-        $datetime = DateUtils::parseDateTime('01/20/2019T13:24:55Z', $timezone);
-        $this->assertEquals("20/01/2019T08:24:55", $datetime->format('d/m/Y\TH:i:s'));
+    public function parseDateTimeProvider(): array {
+        return [
+            ['01/20/2019T13:24:55Z', 'Etc/GMT+5', 'd/m/Y\TH:i:s', '20/01/2019T08:24:55'],
+            // max time value
+            ['9999/12/31T13:24:55Z', 'Etc/GMT-5', 'd/m/Y\TH:i:s', '01/02/9999T00:00:00'],
+            // test timezones with DST
+            ['2020-06-01 12:00:00-04', 'America/New_York', 'Y-m-d H:i:s', '2020-06-01 12:00:00'],
+            ['2020-12-01 12:00:00-05', 'America/New_York', 'Y-m-d H:i:s', '2020-12-01 12:00:00'],
+            ['2020-06-01 12:00:00-04', 'America/Los_Angeles', 'Y-m-d H:i:s', '2020-06-01 09:00:00'],
+            ['2020-12-01 12:00:00-05', 'America/Los_Angeles', 'Y-m-d H:i:s', '2020-12-01 09:00:00'],
+        ];
+    }
+
+    /**
+     * @dataProvider parseDateTimeProvider
+     */
+    public function testParseDateTimeString(string $input, string $timezoneString, string $format, string $expected): void {
+        $timezone = new \DateTimeZone($timezoneString);
+        $datetime = DateUtils::parseDateTime($input, $timezone);
+        $this->assertEquals($expected, $datetime->format($format));
         $this->assertEquals($timezone, $datetime->getTimezone());
     }

@@ -137,95 +153,6 @@ class DateUtilsTester extends \PHPUnit\Framework\TestCase {
         $this->assertEquals('2019-01-20 13:24:55', $actual);
     }

-    /**
-     * Time stamp with -04 offset into America/New_York
-     * Date format: YYYY-MM-DD HH:MM:SS
-     *
-     * Time stamp: 2020-06-01 12:00:00-04
-     * UTC time: 2020-06-01 16:00:00
-     *
-     * Expected: 2020-06-01 12:00:00
-     */
-    public function testParseDateTimeNYOne() {
-        $format = 'Y-m-d H:i:s';
-        $time_zone = new \DateTimeZone('America/New_York');
-
-        $time_stamp = '2020-06-01 12:00:00-04';
-        $date_time = DateUtils::parseDateTime($time_stamp, $time_zone);
-
-        $expected_string = '2020-06-01 12:00:00';
-        $captured_string = $date_time->format($format);
-
-        $this->assertEquals($expected_string, $captured_string);
-    }
-
-    /**
-     * Time stamp with -05 offset into America/New_York
-     * Date format: YYYY-MM-DD HH:MM:SS
-     *
-     * Time stamp: 2020-12-01 12:00:00-05
-     * UTC time: 2020-12-01 17:00:00
-     *
-     * Expected: 2020-12-01 12:00:00
-     */
-    public function testParseDateTimeNYTwo() {
-        $format = 'Y-m-d H:i:s';
-        $time_zone = new \DateTimeZone('America/New_York');
-
-        $time_stamp = '2020-12-01 12:00:00-05';
-        $date_time = DateUtils::parseDateTime($time_stamp, $time_zone);
-
-        $expected_string = '2020-12-01 12:00:00';
-        $captured_string = $date_time->format($format);
-
-        $this->assertEquals($expected_string, $captured_string);
-    }
-
-    /**
-     * Time stamp with -04 offset into America/Los_Angeles
-     * Date format: YYYY-MM-DD HH:MM:SS
-     *
-     * Time stamp: 2020-06-01 12:00:00-04
-     * UTC time: 2020-06-01 16:00:00
-     *
-     * Expected: 2020-06-01 09:00:00
-     */
-    public function testParseDateTimeLAOne() {
-        $format = 'Y-m-d H:i:s';
-        $time_zone = new \DateTimeZone('America/Los_Angeles');
-
-        $time_stamp = '2020-06-01 12:00:00-04';
-        $date_time = DateUtils::parseDateTime($time_stamp, $time_zone);
-
-        $expected_string = '2020-06-01 09:00:00';
-        $captured_string = $date_time->format($format);
-
-        $this->assertEquals($expected_string, $captured_string);
-    }
-
-    /**
-     * Time stamp with -05 offset into America/Los_Angeles
-     * Date format: YYYY-MM-DD HH:MM:SS
-     *
-     * Time stamp: 2020-12-01 12:00:00-05
-     * UTC time: 2020-12-01 17:00:00
-     *
-     * Expected: 2020-12-01 09:00:00
-     */
-    public function testParseDateTimeLATwo() {
-        $format = 'Y-m-d H:i:s';
-        $time_zone = new \DateTimeZone('America/Los_Angeles');
-
-        $time_stamp = '2020-12-01 12:00:00-05';
-        $date_time = DateUtils::parseDateTime($time_stamp, $time_zone);
-
-        $expected_string = '2020-12-01 09:00:00';
-        $captured_string = $date_time->format($format);
-
-        $this->assertEquals($expected_string, $captured_string);
-    }
-
-
     public function testGetServerTime() {
         $core = new Core();
         $config = new Config($core);
```

**Pull request merged**